pkg/apis/sparkscheduler/v1beta1/conversion_resource_reservation_test.go

```diff
@@ -0,0 +1,197 @@
+		t.Fatalf("Conversion from v1Beta1 to v1Beta2 failed with err: %s", err)
+	}
+	compareV1Beta2ResourceReservationSpecs(t, &v1Beta2ReservationWithGPU.Spec, &v1beta2ResConverted.Spec)
+	require.Equal(t, v1Beta2ReservationWithGPU.Status, v1beta2ResConverted.Status)
+	require.Empty(t, v1beta2ResConverted.ObjectMeta.Annotations)
+}
+
+func TestConversionFromV1Beta1ToV1Beta2WithoutGPUs(t *testing.T) {
+	// We expect the v1Beta2 struct to only have CPU and Memory set after fetching the values from the v1beta1
+	// resource reservation spec directly
+	var v1beta2ResConverted v1beta2.ResourceReservation
+	err := v1Beta1ReservationWithoutGPU.ConvertTo(&v1beta2ResConverted)
+	if err != nil {
+		t.Fatalf("Conversion from v1Beta1 to v1Beta2 failed with err: %s", err)
+	}
+	compareV1Beta2ResourceReservationSpecs(t, &v1Beta2ReservationWithoutGPU.Spec, &v1beta2ResConverted.Spec)
+	require.Equal(t, v1Beta2ReservationWithoutGPU.Status, v1beta2ResConverted.Status)
+	require.Empty(t, v1beta2ResConverted.ObjectMeta.Annotations)
+}
+
+func TestConversionFromV2ToV1ToV2AfterChangingValuesInV1(t *testing.T) {
+	// We expect the final v1Beta2 struct to contain the changes that we made to the v1 struct as well as the gpu information
+	// from the initial v2 object.
+
+	// Convert to v1beta1
+	var v1beta1ResConverted ResourceReservation
+	err := v1beta1ResConverted.ConvertFrom(&v1Beta2ReservationWithGPU)
+	if err != nil {
+		t.Fatalf("Conversion from v1Beta2 to v1Beta1 failed with err: %s", err)
+	}
+
+	// Change the reservation object
+	reservation, _ := v1beta1ResConverted.Spec.Reservations["driver"]
+	reservation.CPU = *resource.NewQuantity(10, resource.DecimalSI)
+	v1beta1ResConverted.Spec.Reservations["driver"] = reservation
+
+	// Convert back to v1beta2
+	var v1beta2ResConverted v1beta2.ResourceReservation
+	err = v1beta1ResConverted.ConvertTo(&v1beta2ResConverted)
+	if err != nil {
+		t.Fatalf("Conversion from v1Beta1 to v1Beta2 failed with err: %s", err)
+	}
+
+	compareV1Beta2ResourceReservationSpecs(t, &v1Beta2ReservationWithGPUAndPreConversionChanges.Spec, &v1beta2ResConverted.Spec)
+	require.Equal(t, v1Beta2ReservationWithGPUAndPreConversionChanges.Status, v1beta2ResConverted.Status)
+	require.Empty(t, v1Beta2ReservationWithGPUAndPreConversionChanges.ObjectMeta.Annotations)
+}
+
+func cacheStringValuesOfReservations(r *v1beta2.ResourceReservationSpec) {
+	// Calling String() caches the string value of the quantity, unmarshalled reservations already have this so we need
+	// to call it for all reservations to get deep equality to be consistent
+	for _, value := range r.Reservations {
+		_ = value.Resources.CPU().String()
+		_ = value.Resources.Memory().String()
+		_ = value.Resources.NvidiaGPU().String()
+	}
+}
+
+func compareV1Beta2ResourceReservationSpecs(t *testing.T, r1 *v1beta2.ResourceReservationSpec, r2 *v1beta2.ResourceReservationSpec) {
+	cacheStringValuesOfReservations(r1)
+	cacheStringValuesOfReservations(r2)
+	require.Equal(t, r1, r2)
+}
```

Comment:
Can you implement the equality check instead

Reply:
You can use `cmp.Equal()` which works in this case because the Quantity implements an `Equal()` method